docs/natspec-format.rst

```diff
@@ -0,0 +1,216 @@
+##############
+NatSpec Format
+##############
+
+Solidity contracts can use a special form of comments to provide rich
+documentation for functions, return variables and more. This special form is
+named the Ethereum Natural Specification Format (NatSpec).
+
+This documentation is segmented into developer-focused messages and end-user-
+facing messages. These messages may be shown to the end user (the human) at the
+time that they will interact with the contract (i.e. sign a transaction).
+
+It is recommended that Solidity contracts are fully annontated using NatSpec for
+all public interfaces (everything in the ABI).
+
+NatSpec includes the formatting for comments that the smart contract author will
+use, and which are understood by the Solidity compiler. Also detailed below is
+output of the Solidity compiler, which extracts these comments into a machine-
+readable format.
+
+.. _header-doc-example:
+
+Documentation Example
+=====================
+
+Documentation is inserted above each ``class``, ``interface`` and
+``function`` using the doxygen notation format.
+
+-  For Solidity you may choose ``///`` for single or multi-line
+   comments, or ``/**`` and ending with ``*/``.
+
+-  For Vyper, use ``"""`` indented to the inner contents with bare
+   comments. See `Vyper
+   documentation <https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#natspec-metadata>`__.
+
+This example shows a contract and a function using all available tags.
+Note: NatSpec currently does NOT apply to public state variables (see
+`solidity#3418 <https://github.com/ethereum/solidity/issues/3418>`__),
+even if they are declared public and therefore do affect ABI. Note:
+NatSpec currently only interprets tags functions if they are external or
+public. You are welcome to use similar comments for your internal and
+private functions, but those will not be parsed.
+
+.. code:: solidity
+
+   pragma solidity ^0.5.6;
+
+   /// @title A simulator for trees
+   /// @author Larry A. Gardner
+   /// @notice You can use this contract for only the most basic simulation
+   /// @dev All function calls are currently implemented without side effects
+   contract Tree {
+       /// @author Mary A. Botanist
+       /// @notice Calculate tree age in years, rounded up, for live trees
+       /// @dev The Alexandr N. Tetearing algorithm could increase precision
+       /// @param rings The number of rings from dendrochronological sample
+       /// @return age in years, rounded up for partial years
+       function age(uint256 rings) external pure returns (uint256) {
+           return rings + 1;
+       }
+   }
+
+.. _header-tags:
+
+Tags
+====
+
+All tags are optional. The following table explains the purpose of each
+NatSpec tag and where it may be used.
```

Comment:
If no tag is given, `@notice` is assumed.

Reply:
Updated in 8233930